pkg/spec/deploymentconfig.go

```diff
@@ -0,0 +1,165 @@
+	if err != nil {
+		return errors.Wrap(err, "Unable to fix services")
+	}
+
+	// fix deployment.VolumeClaims
+	deployment.VolumeClaims, err = fixVolumeClaims(deployment.VolumeClaims, deployment.Name)
+	if err != nil {
+		return errors.Wrap(err, "Unable to fix persistentVolume")
+	}
+
+	// fix deployment.configMaps
+	deployment.ConfigMaps, err = fixConfigMaps(deployment.ConfigMaps, deployment.Name)
+	if err != nil {
+		return errors.Wrap(err, "unable to fix configMaps")
+	}
+
+	deployment.Containers, err = fixContainers(deployment.Containers, deployment.Name)
+	if err != nil {
+		return errors.Wrap(err, "unable to fix containers")
+	}
+
+	deployment.InitContainers, err = fixContainers(deployment.InitContainers, deployment.Name)
+	if err != nil {
+		return errors.Wrap(err, "unable to fix init-containers")
+	}
+
+	deployment.Secrets, err = fixSecrets(deployment.Secrets, deployment.Name)
+	if err != nil {
+		return errors.Wrap(err, "unable to fix secrets")
+	}
+
+	return nil
+}
+
+func (deployment *DeploymentConfigSpecMod) Transform() ([]runtime.Object, []string, error) {
+
+	// Create Kubernetes objects (since OpenShift uses Kubernetes underneath, no need to refactor
+	// this portion
+	runtimeObjects, extraResources, err := deployment.CreateK8sObjects()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create Kubernetes objects")
+	}
+
+	// Set appropriate GVK BEFORE adding DeploymentConfig controller
+	// as OpenShift controllers are not available in the Kubernetes controller / setGVK check
+	scheme, err := GetScheme()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to get scheme")
+	}
+
+	// Set's the appropriate GVK
+	for _, runtimeObject := range runtimeObjects {
+		if err := SetGVK(runtimeObject, scheme); err != nil {
+			return nil, nil, errors.Wrap(err, "unable to set Group, Version and Kind for generated Kubernetes resources")
+		}
+	}
+
+	// Create the DeploymentConfig controller!
+	deploy, err := deployment.createDeploymentConfigController()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create Kubernetes Deployment controller")
+	}
+
+	// adding controller objects
+	// deployment will be nil if no deployment is generated and no error occurs,
+	// so we only need to append this when a legit deployment resource is returned
+	if deploy != nil {
+		runtimeObjects = append(runtimeObjects, deploy)
+		log.Debugf("deployment: %s, deployment: %s\n", deploy.Name, spew.Sprint(deployment))
+	}
+
+	if len(runtimeObjects) == 0 {
+		return nil, nil, errors.New("No runtime objects created, possibly because not enough input data was passed")
+	}
+
+	return runtimeObjects, extraResources, nil
+}
+
+// Created the OpenShift DeploymentConfig controller
+func (deployment *DeploymentConfigSpecMod) createDeploymentConfigController() (*os_deploy_v1.DeploymentConfig, error) {
+
+	// Set replicas to 1 if not set (0)
+	replicas := deployment.Replicas
+	if replicas == 0 {
+		replicas = 1
+	}
+
```

Comment:
This method is missing all the checks which are in CreateK8sController() for Kubernetes Deployment.
e.g. -
```golang
	// We need to error out if both, deployment.PodSpec and deployment.DeploymentSpec are empty
	if deployment.isDeploymentSpecPodSpecEmpty() {
		log.Debug("Both, deployment.PodSpec and deployment.DeploymentSpec are empty, not enough data to create a deployment.")
		return nil, nil
	}

	// We are merging whole DeploymentSpec with PodSpec.
	// This means that someone could specify containers in template.spec and also in top level PodSpec.
	// This stupid check is supposed to make sure that only one of them set.
	// TODO: merge DeploymentSpec.Template.Spec and top level PodSpec
	if deployment.isMultiplePodSpecSpecified() {
		return nil, fmt.Errorf("Pod can't be specfied in two places. Use top level PodSpec or template.spec (DeploymentSpec.Template.Spec) not both")
	}

	deploymentSpec := deployment.DeploymentSpec

	// top level PodSpec is not empty, use it for deployment template
	// we already know that if deployment.PodSpec is not empty deployment.DeploymentSpec.Template.Spec is empty
	if !reflect.DeepEqual(deployment.PodSpec, api_v1.PodSpec{}) {
		deploymentSpec.Template.Spec = deployment.PodSpec
	}

	// TODO: check if this wasn't set by user, in that case we shouldn't overwrite it
	deploymentSpec.Template.ObjectMeta.Name = deployment.Name

	// TODO: merge with already existing labels and avoid duplication
	deploymentSpec.Template.ObjectMeta.Labels = deployment.Labels
```

I think it should be in the scope of this PR to abstract that code and reuse it in both the places.

Reply:
@containscafeine IMO we should add this after since OpenShift does things a bit different for pod checking in this PR.

Could we abstract this and refactor in a new PR (since we copy and paste most of this between deployment.go, deploymentconfig.go and job.go)...